README.md
# koa-pagination

`koa-pagination` is a middleware to handle [Range Pagination Headers](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html) using `Range` & `Content-Range` entity-headers.

## Status

[![npm version][npm-image]][npm-url] [![build status][travis-image]][travis-url]

## Installation

Install the package via `yarn`:

```sh
❯ yarn add koa-pagination
```

or via `npm`:

```sh
❯ npm install koa-pagination --save
```

### Configuration

The middleware can be configured with the following parameters:

- `allowAll`: Whether to accept `*` as range-specifier.
- `maximum`: Maximum number of items allowed per page (`50` by default).
- `unit`: Accepted range unit (`items` by default).

You can change the defaults by doing:

```javascript
middleware({
  allowAll: true,
  maximum: 100,
  unit: 'bytes'
});
```

## Usage

```javascript
const { middleware } = require('koa-pagination');
const Koa = require('koa');

const app = new Koa();

app.get('/', middleware(), async ctx => {
  // `paginate` middleware will inject a `pagination` object in the `koa` context,
  // which will allow you to use the `pagination.offset` and `pagination.limit`
  // in your data retrieval methods.
  ctx.body = await foobar.getData({
    limit: ctx.pagination.limit,
    offset: ctx.pagination.offset
  });

  // This is needed in order to expose the length in `Content-Range` header.
  ctx.pagination.length = foobar.count();
});

app.listen(3000);
```

### Request

You can provide the `Range` header specifying the items you want to retrieve. For instance to retrieve the first 5 elements:

```javascript
'Range: items=0-4'
```

You can also provide `*` at the end of the range in order to retrieve the all of the available items:

```javascript
'Range: items=0-*'
```

### Response

The first example will generate a response with the following headers:

```javascript
'Accepted-Ranges': 'items'
'Content-Range: items 0-4/*'
```

The `*` in the `Content-Range` header will be replaced with the total number of items provided in the `length` variable.

#### Codes

Code | Reason
---- | -------------------------------------------------------------------------
200  | `Range` header has not been provided.
206  | `Range` header is valid.
412  | `Range` header is malformed.
416  | `Range` header is invalid.
500  | Incorrect middleware configuration or unexpected value inside middleware.

## Tests

```sh
❯ yarn test
```

## Release

```sh
❯ npm version [<new version> | major | minor | patch] -m "Release %s"
```

[npm-image]: https://img.shields.io/npm/v/koa-pagination.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-pagination
[travis-image]: https://img.shields.io/travis/uphold/koa-pagination.svg?style=flat-square
[travis-url]: https://travis-ci.org/uphold/koa-pagination


CHANGELOG.md
# Changelog

## [v4.0.0](https://github.com/uphold/koa-pagination/releases/tag/v4.0.0) (2019-10-15)
- Add github-changelog-generator as dev-dependency [\#59](https://github.com/uphold/koa-pagination/pull/59) ([eduardoconceicao](https://github.com/eduardoconceicao))
- Update sshpk to 1.16.1 [\#58](https://github.com/uphold/koa-pagination/pull/58) ([eduardoconceicao](https://github.com/eduardoconceicao))
- Update dependencies [\#57](https://github.com/uphold/koa-pagination/pull/57) ([eduardoconceicao](https://github.com/eduardoconceicao))
- Remove is-safe-integer dependency [\#54](https://github.com/uphold/koa-pagination/pull/54) ([nunofgs](https://github.com/nunofgs))
- Update standard-http-error@2.0.1 [\#55](https://github.com/uphold/koa-pagination/pull/55) ([nunofgs](https://github.com/nunofgs))
- Expose `Accept-Ranges` header [\#52](https://github.com/uphold/koa-pagination/pull/52) ([nunofgs](https://github.com/nunofgs))
- Fix invalid range unit override [\#51](https://github.com/uphold/koa-pagination/pull/51) ([nunofgs](https://github.com/nunofgs))
- Do not overwrite status if it is not successful [\#50](https://github.com/uphold/koa-pagination/pull/50) ([nunofgs](https://github.com/nunofgs))
- Update ownership [\#46](https://github.com/uphold/koa-pagination/pull/46) ([ruimarinho](https://github.com/ruimarinho))

## [v3.0.0](https://github.com/uphold/koa-pagination/releases/tag/v3.0.0) (2017-03-30)
- Move source to /src [\#45](https://github.com/uphold/koa-pagination/pull/45) ([kurayama](https://github.com/kurayama))
- Improve README [\#44](https://github.com/uphold/koa-pagination/pull/44) ([kurayama](https://github.com/kurayama))
- Replace default export by named middleware and errors [\#43](https://github.com/uphold/koa-pagination/pull/43) ([nunofgs](https://github.com/nunofgs))
- Add .npmignore [\#42](https://github.com/uphold/koa-pagination/pull/42) ([ruimarinho](https://github.com/ruimarinho))
- Update koa@2 [\#40](https://github.com/uphold/koa-pagination/pull/40) ([kurayama](https://github.com/kurayama))
- Add LICENSE [\#41](https://github.com/uphold/koa-pagination/pull/41) ([ruimarinho](https://github.com/ruimarinho))
- Update README [\#38](https://github.com/uphold/koa-pagination/pull/38) ([franciscocardoso](https://github.com/franciscocardoso))
- Add `allowAll` parameter to disable asterisk [\#37](https://github.com/uphold/koa-pagination/pull/37) ([nunofgs](https://github.com/nunofgs))

## [v2.0.1](https://github.com/uphold/koa-pagination/releases/tag/v2.0.1) (2016-05-18)
- Update range-specifier-parser@1.0.2 [\#35](https://github.com/uphold/koa-pagination/pull/35) ([nunofgs](https://github.com/nunofgs))
- Add asterisk instructions to readme [\#34](https://github.com/uphold/koa-pagination/pull/34) ([nunofgs](https://github.com/nunofgs))

## [v2.0.0](https://github.com/uphold/koa-pagination/releases/tag/v2.0.0) (2016-05-18)
- Add asterisk support [\#32](https://github.com/uphold/koa-pagination/pull/32) ([nunofgs](https://github.com/nunofgs))
- Refactor to es6 standards [\#31](https://github.com/uphold/koa-pagination/pull/31) ([nunofgs](https://github.com/nunofgs))
- Update dependencies [\#30](https://github.com/uphold/koa-pagination/pull/30) ([nunofgs](https://github.com/nunofgs))

## [v1.1.1](https://github.com/uphold/koa-pagination/releases/tag/v1.1.1) (2016-05-04)
- Update README [\#29](https://github.com/uphold/koa-pagination/pull/29) ([franciscocardoso](https://github.com/franciscocardoso))

## [v1.1.0](https://github.com/uphold/koa-pagination/releases/tag/v1.1.0) (2016-05-04)

## [v0.1.1](https://github.com/uphold/koa-pagination/releases/tag/v0.1.1) (2016-05-04)

## [v0.1.2](https://github.com/uphold/koa-pagination/releases/tag/v0.1.2) (2016-05-04)

## [v0.1.3](https://github.com/uphold/koa-pagination/releases/tag/v0.1.3) (2016-05-04)

## [v0.2.0](https://github.com/uphold/koa-pagination/releases/tag/v0.2.0) (2016-05-04)

## [v0.3.0](https://github.com/uphold/koa-pagination/releases/tag/v0.3.0) (2016-05-04)

## [v0.3.1](https://github.com/uphold/koa-pagination/releases/tag/v0.3.1) (2016-05-04)

## [v1.0.0](https://github.com/uphold/koa-pagination/releases/tag/v1.0.0) (2016-05-04)

## [v0.1.0](https://github.com/uphold/koa-pagination/releases/tag/v0.1.0) (2016-05-04)
- Update changelog and version scripts [\#28](https://github.com/uphold/koa-pagination/pull/28) ([franciscocardoso](https://github.com/franciscocardoso))
- Expose `range-unit` [\#26](https://github.com/uphold/koa-pagination/pull/26) ([franciscocardoso](https://github.com/franciscocardoso))
- Improve error message [\#27](https://github.com/uphold/koa-pagination/pull/27) ([ruimarinho](https://github.com/ruimarinho))
- Remove usage of `chai` [\#25](https://github.com/uphold/koa-pagination/pull/25) ([franciscocardoso](https://github.com/franciscocardoso))
- Replace `create-error` with `standard-http-error` [\#23](https://github.com/uphold/koa-pagination/pull/23) ([franciscocardoso](https://github.com/franciscocardoso))
- Prevent usage of unsafe numbers [\#24](https://github.com/uphold/koa-pagination/pull/24) ([franciscocardoso](https://github.com/franciscocardoso))
- Fix readme headers example [\#21](https://github.com/uphold/koa-pagination/pull/21) ([nunorafaelrocha](https://github.com/nunorafaelrocha))
- Change default unit to `items` [\#20](https://github.com/uphold/koa-pagination/pull/20) ([fixe](https://github.com/fixe))
- Allow specifying `unit` option [\#19](https://github.com/uphold/koa-pagination/pull/19) ([fixe](https://github.com/fixe))
- Update response status [\#18](https://github.com/uphold/koa-pagination/pull/18) ([franciscocardoso](https://github.com/franciscocardoso))
- Add codeclimate `gpa` and `coverage` badges [\#13](https://github.com/uphold/koa-pagination/pull/13) ([franciscocardoso](https://github.com/franciscocardoso))
- Add npm badge [\#14](https://github.com/uphold/koa-pagination/pull/14) ([fixe](https://github.com/fixe))

src/index.js
'use strict';

/**
 * Module dependencies.
 */

const contentRangeFormat = require('http-content-range-format');
const errors = require('./errors');
const rangeSpecifierParser = require('range-specifier-parser').default;

/**
 * Instances.
 */

const { InvalidConfigurationError, MalformedRangeError, RangeNotSatisfiableError } = errors;

/**
 * Paginate middleware.
 */

function middleware({ allowAll = true, maximum = 50, unit = 'items' } = {}) {
  return async function paginate(ctx, next) {
    let first = 0;
    let last = maximum;
    let limit = '*';

    // Prevent invalid `maximum` value configuration.
    if (!isFinite(maximum) || !Number.isSafeInteger(maximum) || maximum <= 0) {
      throw new InvalidConfigurationError();
    }

    // Handle `Range` header.
    if (ctx.get('Range')) {
      const range = rangeSpecifierParser(ctx.get('Range'));

      if (range === -1) {
        throw new RangeNotSatisfiableError();
      }

      if (range === -2) {
        throw new MalformedRangeError();
      }

      if (range.unit !== unit) {
        throw new MalformedRangeError();
      }

      // Update `limit`, `offset` values.
      first = range.first;
      last = range.last;

      if (!allowAll && last === '*') {
        throw new RangeNotSatisfiableError();
      }

      if (!Number.isSafeInteger(first) || (last !== '*' && !Number.isSafeInteger(last))) {
        throw new RangeNotSatisfiableError();
      }
    }

    if (Number.isSafeInteger(last)) {
      // Prevent pages to be longer than allowed.
      if (last - first + 1 > maximum) {
        last = first + maximum - 1;
      }

      // Calculate limit in the specified range.
      limit = last - first + 1;
    }

    // Set pagination object on context.
    ctx.pagination = {
      limit,
      offset: first,
      unit
    };

    await next();

    const { length } = ctx.pagination;

    // Prevent nonexistent pages.
    if (first > length - 1 && length > 0) {
      throw new RangeNotSatisfiableError();
    }

    // Set the calculated `last` value.
    if (last === '*') {
      last = length;
    }

    // Fix `last` value if `length` is lower.
    if (last + 1 > length && length !== 0) {
      last = length - 1;
    }

    // Set `byte-range-spec` to undefined value - `*`.
    if (length === 0) {
      first = undefined;
      last = undefined;
    }

    // Set response headers based on available units.
    ctx.set('Accept-Ranges', unit);
    ctx.set('Content-Range', contentRangeFormat({ first, last, length, unit }));

    // Allow non-successful status codes.
    if (ctx.status < 200 || ctx.status > 300) {
      return;
    }

    // Set the response as `Partial Content`.
    ctx.status = 206;
  };
}

/**
 * Exports.
 */

module.exports = { errors, middleware };
